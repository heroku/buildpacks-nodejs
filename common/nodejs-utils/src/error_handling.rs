use bullet_stream::{style, Print};
use indoc::formatdoc;
use std::fmt::{Display, Formatter};
use std::path::Path;

pub fn on_framework_error<E>(buildpack_name: &str, issues_url: &str, error: &E) -> ErrorMessage
where
    E: Display,
{
    let issues_url = style::url(issues_url);
    error_message()
        .error_type(ErrorType::Framework)
        .header(format!("{buildpack_name} internal error"))
        .body(formatdoc! {"
            The framework used by this buildpack encountered an unexpected error.

            If you canâ€™t deploy to Heroku due to this issue, check the official Heroku Status page at \
            status.heroku.com for any ongoing incidents. After all incidents resolve, retry your build.

            Use the debug information above to troubleshoot and retry your build. If you think you found a \
            bug in the buildpack, reproduce the issue locally with a minimal example and file an issue here:
            {issues_url}
        "})
        .debug_info(error.to_string())
        .issues_url(issues_url)
        .create()
}

#[bon::builder(finish_fn = create, on(String, into), state_mod(vis = "pub"))]
#[allow(clippy::needless_pass_by_value)]
pub fn error_message(
    header: String,
    body: String,
    error_type: ErrorType,
    issues_url: String,
    debug_info: Option<String>,
) -> ErrorMessage {
    let mut message_parts = vec![header.trim().to_string(), body.trim().to_string()];
    let issues_url = style::url(issues_url);
    let pack = style::value("pack");
    let pack_url =
        style::url("https://buildpacks.io/docs/for-platform-operators/how-to/integrate-ci/pack/");

    match error_type {
        ErrorType::Framework => {}
        ErrorType::Internal => {
            message_parts.push(formatdoc! { "
                The causes for this error are unknown. We do not have suggestions for diagnosis or a \
                workaround at this time. You can help our understanding by sharing your buildpack log \
                and a description of the issue at:
                {issues_url}

                If you're able to reproduce the problem with an example application and the {pack} \
                build tool ({pack_url}), adding that information to the discussion will also help. Once \
                we have more information around the causes of this error we may update this message.
            "});
        }
        ErrorType::UserFacing(suggest_retry_build, suggest_submit_issue) => {
            if let SuggestRetryBuild::Yes = suggest_retry_build {
                message_parts.push(
                    formatdoc! { "
                        Use the debug information above to troubleshoot and retry your build.
                    "}
                    .trim()
                    .to_string(),
                );
            }

            if let SuggestSubmitIssue::Yes = suggest_submit_issue {
                message_parts.push(formatdoc! { "
                    If the issue persists and you think you found a bug in the buildpack, reproduce \
                    the issue locally with a minimal example. Open an issue in the buildpack's GitHub \
                    repository and include the details here:
                    {issues_url}
                "}.trim().to_string());
            }
        }
    }

    let message = message_parts.join("\n\n");

    ErrorMessage {
        debug_info,
        message,
    }
}

pub fn file_value(value: impl AsRef<Path>) -> String {
    style::value(value.as_ref().to_string_lossy())
}

#[derive(Debug)]
pub struct ErrorMessage {
    debug_info: Option<String>,
    message: String,
}

impl Display for ErrorMessage {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut log = Print::new(vec![]).without_header();
        if let Some(debug_info) = &self.debug_info {
            log = log
                .bullet(style::important("Debug Info:"))
                .sub_bullet(debug_info)
                .done();
        }
        let output = log.error(&self.message);
        write!(f, "{}", String::from_utf8_lossy(&output))
    }
}

#[derive(Debug, PartialEq)]
pub enum ErrorType {
    Framework,
    Internal,
    UserFacing(SuggestRetryBuild, SuggestSubmitIssue),
}

#[derive(Debug, PartialEq)]
pub enum SuggestRetryBuild {
    Yes,
    No,
}

#[derive(Debug, PartialEq)]
pub enum SuggestSubmitIssue {
    Yes,
    No,
}
