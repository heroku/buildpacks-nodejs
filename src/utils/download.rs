use crate::utils;
use crate::utils::async_runtime::ASYNC_RUNTIME;
use async_compression::tokio::bufread::GzipDecoder;
use bullet_stream::global::print;
use futures::StreamExt;
use futures::stream::TryStreamExt;
use sha2::{Digest, Sha256};
use std::io;
use std::path::{Path, PathBuf};
use std::time::Duration;
use tokio::fs::File as AsyncFile;
use tokio::io::{AsyncWriteExt, BufReader as AsyncBufReader, copy as async_copy, sink};
use tokio_util::io::{InspectReader, StreamReader};

pub(crate) fn download_sync<'a, T>(downloader: T) -> Result<(), DownloaderError>
where
    T: Downloader<'a>,
{
    ASYNC_RUNTIME.block_on(async { download(downloader).await })
}

#[allow(clippy::too_many_lines)]
pub(crate) async fn download<'a, T>(downloader: T) -> Result<(), DownloaderError>
where
    T: Downloader<'a>,
{
    let response = utils::http::get(downloader.source_url())
        .connect_timeout(downloader.connect_timeout())
        .read_timeout(downloader.read_timeout())
        .max_retries(downloader.max_retries())
        .call()
        .await
        .map_err(|e| DownloaderError::Request {
            url: downloader.source_url().to_string(),
            source: e,
        })?;

    let timer = print::sub_start_timer("Downloading");

    let checksum_type = downloader.checksum_validator();

    #[allow(clippy::manual_map)]
    let mut hasher = match checksum_type {
        Some(ChecksumValidator::Sha256(_)) => Some(Sha256::new()),
        None => None,
    };

    // ensure the writer is dropped in this scope so the InspectReader is dropped and is no longer borrowing the hash digest
    {
        let mut inspectable_reader = AsyncBufReader::new(
            // the inspect reader lets us pipe this decompressed output to both the output file and the hash digest
            InspectReader::new(
                StreamReader::new(response.bytes_stream().map_err(io::Error::other)),
                |bytes| {
                    if let Some(hasher) = &mut hasher {
                        hasher.update(bytes);
                    }
                },
            ),
        );

        let create_write_error = |e| DownloaderError::Write {
            url: downloader.source_url().to_string(),
            destination: downloader.destination().to_path_buf(),
            source: e,
        };

        match downloader.extractor() {
            Some(Extractor::Gzip(options)) => {
                let mut reader = GzipDecoder::new(inspectable_reader);
                // Enable/Disable support for multistream gz files. In this mode, the reader expects the input to
                // be a sequence of individually gzipped data streams, each with its own header and trailer,
                // ending at EOF. This is standard behavior for gzip readers.
                reader.multiple_members(options.multiple_members);
                let mut tar_archive = tokio_tar::Archive::new(reader);
                let mut entries = tar_archive.entries().map_err(create_write_error)?;
                while let Some(entry) = entries.next().await {
                    let mut entry = entry.map_err(create_write_error)?;
                    let path = entry.path().map_err(create_write_error)?;

                    // Get the path components
                    let path_components: Vec<_> = path.components().collect();

                    // Skip if we don't have enough components after stripping
                    if path_components.len() <= options.strip_components {
                        async_copy(&mut entry, &mut sink())
                            .await
                            .map_err(create_write_error)?;
                        continue;
                    }

                    // Skip if the path contains '..' or is absolute
                    if path_components.iter().any(|c| {
                        matches!(
                            c,
                            std::path::Component::ParentDir | std::path::Component::RootDir
                        )
                    }) {
                        async_copy(&mut entry, &mut sink())
                            .await
                            .map_err(create_write_error)?;
                        continue;
                    }

                    // Build the stripped path
                    let stripped_path: PathBuf = path_components
                        .into_iter()
                        .skip(options.strip_components)
                        .collect();

                    // Skip empty paths
                    if stripped_path.components().count() == 0 {
                        async_copy(&mut entry, &mut sink())
                            .await
                            .map_err(create_write_error)?;
                        continue;
                    }

                    // Skip excluded paths
                    let exclude = &options.exclude;
                    if exclude(&stripped_path) {
                        async_copy(&mut entry, &mut sink())
                            .await
                            .map_err(create_write_error)?;
                        continue;
                    }

                    let entry_type = entry.header().entry_type();
                    let dest_path = downloader.destination().join(&stripped_path);

                    // Create parent directories for regular files and symlinks
                    if (entry_type.is_file()
                        || entry_type.is_symlink()
                        || entry_type.is_hard_link())
                        && let Some(parent) = dest_path.parent()
                    {
                        tokio::fs::create_dir_all(parent)
                            .await
                            .map_err(create_write_error)?;
                    }

                    entry.unpack(dest_path).await.map_err(create_write_error)?;
                }
            }
            None => {
                let mut writer = AsyncFile::create(downloader.destination())
                    .await
                    .map_err(create_write_error)?;
                async_copy(&mut inspectable_reader, &mut writer)
                    .await
                    .map_err(create_write_error)?;
                writer.flush().await.map_err(create_write_error)?;
            }
        }
    }

    timer.done();

    if let Some(hasher) = hasher.take() {
        match checksum_type {
            Some(ChecksumValidator::Sha256(expected_value)) => {
                let actual_checksum = hasher.finalize();
                if expected_value != actual_checksum.to_vec() {
                    Err(DownloaderError::ChecksumMismatch {
                        url: downloader.source_url().to_string(),
                        expected_checksum: hex::encode(expected_value),
                        actual_checksum: format!("{actual_checksum:x}"),
                    })?;
                }
            }
            None => {}
        }
    }

    Ok(())
}

pub(crate) trait Downloader<'a> {
    fn source_url(&self) -> &str;

    fn destination(&self) -> &Path;

    fn checksum_validator(&self) -> Option<ChecksumValidator<'a>> {
        None
    }

    fn extractor(&self) -> Option<Extractor> {
        None
    }

    fn connect_timeout(&self) -> Duration {
        crate::utils::http::DEFAULT_CONNECT_TIMEOUT
    }

    fn read_timeout(&self) -> Duration {
        crate::utils::http::DEFAULT_READ_TIMEOUT
    }

    fn max_retries(&self) -> u32 {
        crate::utils::http::DEFAULT_RETRIES
    }
}

#[derive(Debug)]
pub(crate) enum Extractor {
    Gzip(GzipOptions),
}

pub(crate) struct GzipOptions {
    pub(crate) multiple_members: bool,
    pub(crate) strip_components: usize,
    pub(crate) exclude: Box<dyn Fn(&Path) -> bool>,
}

impl std::fmt::Debug for GzipOptions {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("GzipOptions")
            .field("multiple_members", &self.multiple_members)
            .field("strip_components", &self.strip_components)
            .field("exclude", &"<closure>")
            .finish()
    }
}

impl Default for GzipOptions {
    fn default() -> Self {
        Self {
            multiple_members: true,
            strip_components: 0,
            exclude: Box::new(|_| false),
        }
    }
}

#[derive(Debug)]
pub(crate) enum ChecksumValidator<'a> {
    Sha256(&'a [u8]),
}

#[derive(Debug)]
pub(crate) enum DownloaderError {
    Request {
        url: String,
        source: utils::http::Error,
    },
    Write {
        url: String,
        destination: PathBuf,
        source: io::Error,
    },
    ChecksumMismatch {
        url: String,
        expected_checksum: String,
        actual_checksum: String,
    },
}
