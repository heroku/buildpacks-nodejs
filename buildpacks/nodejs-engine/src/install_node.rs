use bullet_stream::global::print;
use bullet_stream::style;
use libcnb::build::BuildContext;
use libcnb::data::layer_name;
use libcnb::layer::{
    CachedLayerDefinition, InvalidMetadataAction, LayerState, RestoredLayerAction,
};
use libherokubuildpack::download::download_file;
use libherokubuildpack::fs::move_directory_contents;
use libherokubuildpack::inventory::artifact::Artifact;
use libherokubuildpack::tar::decompress_tarball;
use serde::{Deserialize, Serialize};
use sha2::digest::Output;
use sha2::{Digest, Sha256};
use std::fs;
use std::io::Read;
use std::path::{Path, PathBuf};
use tempfile::NamedTempFile;

use heroku_nodejs_utils::vrs::Version;

use crate::{NodeJsEngineBuildpack, NodeJsEngineBuildpackError};

pub(crate) fn install_node(
    context: &BuildContext<NodeJsEngineBuildpack>,
    distribution_artifact: &Artifact<Version, Sha256, Option<()>>,
) -> Result<(), libcnb::Error<NodeJsEngineBuildpackError>> {
    print::bullet("Installing Node.js distribution");

    let new_metadata = DistLayerMetadata {
        artifact: distribution_artifact.clone(),
        layer_version: LAYER_VERSION.to_string(),
    };

    let distribution_layer = context.cached_layer(
        layer_name!("dist"),
        CachedLayerDefinition {
            build: true,
            launch: true,
            invalid_metadata_action: &|_| InvalidMetadataAction::DeleteLayer,
            restored_layer_action: &|old_metadata: &DistLayerMetadata, _| {
                if old_metadata == &new_metadata {
                    RestoredLayerAction::KeepLayer
                } else {
                    RestoredLayerAction::DeleteLayer
                }
            },
        },
    )?;

    let version_tag = format!(
        "{} ({}-{})",
        distribution_artifact.version, distribution_artifact.os, distribution_artifact.arch
    );
    match distribution_layer.state {
        LayerState::Restored { .. } => {
            print::sub_bullet(format!("Reusing Node.js {version_tag}"));
        }
        LayerState::Empty { .. } => {
            distribution_layer.write_metadata(new_metadata)?;

            let node_tgz = NamedTempFile::new().map_err(DistLayerError::TempFile)?;

            let timer = print::sub_start_timer(format!(
                "Downloading Node.js {} from {}",
                style::value(&version_tag),
                style::url(&distribution_artifact.url)
            ));
            download_file(&distribution_artifact.url, node_tgz.path()).map_err(|e| {
                DistLayerError::Download {
                    src_url: distribution_artifact.url.clone(),
                    dst_path: node_tgz.path().to_path_buf(),
                    source: e,
                }
            })?;
            let _ = timer.done();

            print::sub_bullet("Verifying checksum");
            let digest = sha256(node_tgz.path()).map_err(|e| DistLayerError::ReadTempFile {
                path: node_tgz.path().to_path_buf(),
                source: e,
            })?;
            if distribution_artifact.checksum.value != digest.to_vec() {
                Err(DistLayerError::ChecksumVerification {
                    url: distribution_artifact.url.clone(),
                    expected: format!(
                        "{:x}",
                        Sha256::digest(&distribution_artifact.checksum.value)
                    ),
                    actual: format!("{digest:x}"),
                })?;
            }

            print::sub_bullet(format!("Extracting Node.js {}", style::value(&version_tag)));
            let node_tgz_path = node_tgz.path().to_path_buf();
            decompress_tarball(&mut node_tgz.into_file(), distribution_layer.path()).map_err(
                |e| DistLayerError::Untar {
                    src_path: node_tgz_path,
                    dst_path: distribution_layer.path().clone(),
                    source: e,
                },
            )?;

            let timer = print::sub_start_timer(format!(
                "Installing Node.js {}",
                style::value(&version_tag)
            ));
            let dist_name = extract_tarball_prefix(&distribution_artifact.url)
                .ok_or_else(|| DistLayerError::TarballPrefix(distribution_artifact.url.clone()))?;
            let dist_path = distribution_layer.path().join(dist_name);
            move_directory_contents(&dist_path, distribution_layer.path()).map_err(|e| {
                DistLayerError::Installation {
                    src_path: dist_path,
                    dst_path: distribution_layer.path(),
                    source: e,
                }
            })?;
            let _ = timer.done();
        }
    }

    Ok(())
}

fn sha256(path: impl AsRef<Path>) -> Result<Output<Sha256>, std::io::Error> {
    let mut file = fs::File::open(path.as_ref())?;
    let mut buffer = [0x00; 10 * 1024];
    let mut sha256: Sha256 = Sha256::default();

    let mut read = file.read(&mut buffer)?;
    while read > 0 {
        Digest::update(&mut sha256, &buffer[..read]);
        read = file.read(&mut buffer)?;
    }

    Ok(sha256.finalize())
}

fn extract_tarball_prefix(url: &str) -> Option<&str> {
    url.rfind('/').and_then(|last_slash| {
        url.rfind(".tar.gz")
            .map(|tar_gz_index| &url[last_slash + 1..tar_gz_index])
    })
}

const LAYER_VERSION: &str = "1";

#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub(crate) struct DistLayerMetadata {
    artifact: Artifact<Version, Sha256, Option<()>>,
    layer_version: String,
}

#[derive(Debug)]
pub(crate) enum DistLayerError {
    TempFile(std::io::Error),
    Download {
        src_url: String,
        dst_path: PathBuf,
        source: libherokubuildpack::download::DownloadError,
    },
    Untar {
        src_path: PathBuf,
        dst_path: PathBuf,
        source: std::io::Error,
    },
    TarballPrefix(String),
    Installation {
        src_path: PathBuf,
        dst_path: PathBuf,
        source: std::io::Error,
    },
    ChecksumVerification {
        url: String,
        expected: String,
        actual: String,
    },
    ReadTempFile {
        path: PathBuf,
        source: std::io::Error,
    },
}

impl From<DistLayerError> for libcnb::Error<NodeJsEngineBuildpackError> {
    fn from(value: DistLayerError) -> Self {
        libcnb::Error::BuildpackError(NodeJsEngineBuildpackError::DistLayer(value))
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use libherokubuildpack::inventory::{
        artifact::{Arch, Os},
        checksum::Checksum,
    };

    use super::*;

    #[test]
    fn dist_metadata_sanity_check() {
        let node_version_22_1_0_linux_arm = Artifact {
            version: Version::from_str("22.1.0").unwrap(),
            os: Os::Linux,
            arch: Arch::Arm64,
            url: "https://nodejs.org/download/release/v22.1.0/node-v22.1.0-linux-arm64.tar.gz"
                .to_string(),
            checksum: "sha256:9c111af1f951e8869615bca3601ce7ab6969374933bdba6397469843b808f222"
                .parse::<Checksum<Sha256>>()
                .unwrap(),
            metadata: None,
        };
        let node_version_22_1_0_linux_amd = Artifact {
            version: Version::from_str("22.1.0").unwrap(),
            os: Os::Linux,
            arch: Arch::Amd64,
            url: "https://nodejs.org/download/release/v22.1.0/node-v22.1.0-linux-x64.tar.gz"
                .to_string(),
            checksum: "sha256:d8ae35a9e2bb0c0c0611ee9bacf564ea51cc8291ace1447f95ee6aeaf4f1d61d"
                .parse::<Checksum<Sha256>>()
                .unwrap(),
            metadata: None,
        };

        // this is a check to ensure that the same node.js artifact doesn't invalidate the cache
        assert_eq!(
            DistLayerMetadata {
                artifact: node_version_22_1_0_linux_arm.clone(),
                layer_version: LAYER_VERSION.to_string()
            },
            DistLayerMetadata {
                artifact: node_version_22_1_0_linux_arm.clone(),
                layer_version: LAYER_VERSION.to_string()
            }
        );

        // this is a check to ensure that a different node.js artifact does invalidate the cache
        assert_ne!(
            DistLayerMetadata {
                artifact: node_version_22_1_0_linux_arm,
                layer_version: LAYER_VERSION.to_string()
            },
            DistLayerMetadata {
                artifact: node_version_22_1_0_linux_amd,
                layer_version: LAYER_VERSION.to_string()
            }
        );
    }

    #[test]
    fn test_metadata_guard() {
        let metadata = DistLayerMetadata {
            artifact: Artifact {
                version: Version::from_str("22.1.0").unwrap(),
                os: Os::Linux,
                arch: Arch::Amd64,
                url: "https://nodejs.org/download/release/v22.1.0/node-v22.1.0-linux-x64.tar.gz"
                    .to_string(),
                checksum: "sha256:d8ae35a9e2bb0c0c0611ee9bacf564ea51cc8291ace1447f95ee6aeaf4f1d61d"
                    .parse::<Checksum<Sha256>>()
                    .unwrap(),
                metadata: None,
            },
            layer_version: LAYER_VERSION.to_string(),
        };
        let actual = toml::to_string(&metadata).unwrap();
        let expected = r#"
layer_version = "1"

[artifact]
version = "22.1.0"
os = "linux"
arch = "amd64"
url = "https://nodejs.org/download/release/v22.1.0/node-v22.1.0-linux-x64.tar.gz"
checksum = "sha256:d8ae35a9e2bb0c0c0611ee9bacf564ea51cc8291ace1447f95ee6aeaf4f1d61d"
"#
        .trim();
        assert_eq!(expected, actual.trim());
        let from_toml: DistLayerMetadata = toml::from_str(&actual).unwrap();
        assert_eq!(metadata, from_toml);
    }
}
