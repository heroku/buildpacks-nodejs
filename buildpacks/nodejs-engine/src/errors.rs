use crate::install_node::DistLayerError;
use crate::NodeJsEngineBuildpackError;
use bullet_stream::style;
use heroku_nodejs_utils::error_handling::error_message_builder::SetIssuesUrl;
use heroku_nodejs_utils::error_handling::ErrorType::{Internal, UserFacing};
use heroku_nodejs_utils::error_handling::{
    file_value, on_framework_error, on_package_json_error, ErrorMessage, ErrorMessageBuilder,
    SuggestRetryBuild, SuggestSubmitIssue,
};
use indoc::formatdoc;
use libcnb::Error;
use libherokubuildpack::download::DownloadError;

const BUILDPACK_NAME: &str = "Heroku Node.js Engine buildpack";

const ISSUES_URL: &str = "https://github.com/heroku/buildpacks-nodejs/issues";

pub(crate) fn on_error(error: Error<NodeJsEngineBuildpackError>) -> ErrorMessage {
    match error {
        Error::BuildpackError(e) => on_buildpack_error(e),
        e => on_framework_error(BUILDPACK_NAME, ISSUES_URL, &e),
    }
}

// Wraps the error_message() builder to preset the issues_url field
fn error_message() -> ErrorMessageBuilder<SetIssuesUrl> {
    heroku_nodejs_utils::error_handling::error_message().issues_url(ISSUES_URL.to_string())
}

fn on_buildpack_error(error: NodeJsEngineBuildpackError) -> ErrorMessage {
    match error {
        NodeJsEngineBuildpackError::InventoryParse(e) => on_inventory_parse_error(&e),
        NodeJsEngineBuildpackError::PackageJson(e) => {
            on_package_json_error(BUILDPACK_NAME, ISSUES_URL, e)
        }
        NodeJsEngineBuildpackError::UnknownVersion(e) => on_unknown_version_error(e),
        NodeJsEngineBuildpackError::DistLayer(e) => on_dist_layer_error(e),
    }
}

fn on_inventory_parse_error(error: &toml::de::Error) -> ErrorMessage {
    let inventory_file = file_value("./inventory.toml");
    error_message()
        .error_type(Internal)
        .header("Couldn't parse Node.js inventory")
        .body(formatdoc! {"
            The {BUILDPACK_NAME} embeds the content of the {inventory_file} found in this buildpack's \
            repository into its binary to look up and resolve Node.js versions but the content isn't \
            valid TOML.
        "})
        .debug_info(error.to_string())
        .create()
}

fn on_unknown_version_error(version: String) -> ErrorMessage {
    let node_releases_url = style::url(format!(
        "https://github.com/nodejs/node/releases?q=v{version}&expanded=true"
    ));
    let inventory_url = style::url("https://github.com/heroku/buildpacks-nodejs/blob/main/buildpacks/nodejs-engine/inventory.toml");
    let version = style::value(version);
    error_message()
        .error_type(UserFacing(SuggestRetryBuild::Yes, SuggestSubmitIssue::Yes))
        .header(format!("Unknown Node.js version: {version}"))
        .body(formatdoc! {"
            The Node.js version provided could not be resolved to a known release in this buildpack's \
            inventory of Node.js releases.
            
            Suggestions:
            - Confirm if this is a valid Node.js release at {node_releases_url}
            - Check if this buildpack includes the requested Node.js version in its inventory file at {inventory_url}     
        "})
        .create()
}

#[allow(clippy::too_many_lines)]
fn on_dist_layer_error(error: DistLayerError) -> ErrorMessage {
    match error {
        DistLayerError::TempFile(e) => error_message()
            .error_type(UserFacing(SuggestRetryBuild::Yes, SuggestSubmitIssue::No))
            .header("Failed to create temporary download file")
            .body(formatdoc! { "
                The {BUILDPACK_NAME} downloads the target Node.js distribution to a temporary file \
                before installation but an unexpected error occurred.
            "})
            .debug_info(e.to_string())
            .create(),

        DistLayerError::Download {
            src_url,
            dst_path,
            source,
        } => {
            let nodejs_status_url = style::url("https://status.nodejs.org/");
            let src_url = style::url(src_url);
            let dst_path = file_value(dst_path);
            match source {
                DownloadError::HttpError(e) => error_message()
                    .error_type(UserFacing(SuggestRetryBuild::Yes, SuggestSubmitIssue::No))
                    .header("Failed to download Node.js distribution")
                    .body(formatdoc! {"
                        A request to download the target Node.js distribution from {src_url} failed unexpectedly. This error can \
                        occur due to an unstable network connection or an issue with the upstream Node.js distribution \
                        server.
                        
                        Suggestions:
                        - Check the status of {nodejs_status_url} for any reported issues.
                        - Confirm the download url ({src_url}) works. 
                    "})
                    .debug_info(e.to_string())
                    .create(),

                DownloadError::IoError(e) => error_message()
                    .error_type(UserFacing(SuggestRetryBuild::Yes, SuggestSubmitIssue::Yes))
                    .header("Failed to write downloaded Node.js distribution")
                    .body(formatdoc! {"
                        An unexpected I/O error occurred while writing the Node.js distribution from {src_url} to \
                        disk at {dst_path}.
                    "})
                    .debug_info(e.to_string())
                    .create()
            }
        }

        DistLayerError::Untar {
            src_path,
            dst_path,
            source,
        } => {
            let src_path = file_value(src_path);
            let dst_path = file_value(dst_path);
            error_message()
                .error_type(UserFacing(SuggestRetryBuild::Yes, SuggestSubmitIssue::Yes))
                .header("Failed to unpack Node.js distribution")
                .body(formatdoc! {"
                    An unexpected I/O error occurred while trying to unpack from {src_path} to {dst_path}.
                "})
                .debug_info(source.to_string())
                .create()
        }

        DistLayerError::TarballPrefix(url) => {
            let url = style::url(url);
            error_message()
                .error_type(Internal)
                .header("Could not determine tarball extraction directory")
                .body(formatdoc! {"
                    An unexpected error occurred while trying to determine the name of the tarball \
                    extraction directory from {url}.
                "})
                .create()
        }

        DistLayerError::Installation {
            src_path,
            dst_path,
            source,
        } => {
            let src_path = file_value(src_path);
            let dst_path = file_value(dst_path);
            error_message()
                .error_type(UserFacing(SuggestRetryBuild::Yes, SuggestSubmitIssue::Yes))
                .header("Failed to copy Node.js distribution contents")
                .body(formatdoc! {"
                    An unexpected I/O error occurred while copying the contents of the Node.js \
                    distribution from {src_path} to the installation directory at {dst_path}. 
                "})
                .debug_info(source.to_string())
                .create()
        }

        DistLayerError::ChecksumVerification {
            url,
            expected,
            actual,
        } => {
            let url = style::url(url);
            let expected = style::value(expected);
            let actual = style::value(actual);
            error_message()
                .error_type(UserFacing(SuggestRetryBuild::Yes, SuggestSubmitIssue::Yes))
                .header("Node.js distribution checksum verification failed")
                .body(formatdoc! {"
                    An error occurred while verifying the checksum of the Node.js distribution from \
                    {url}. This error can occur due to an issue with the upstream Node.js download server.

                    Checksum:
                    - Expected: {expected}
                    - Actual: {actual}    
                "})
                .create()
        }

        DistLayerError::ReadTempFile { path, source } => {
            let path = file_value(path);
            error_message()
                .error_type(UserFacing(SuggestRetryBuild::Yes, SuggestSubmitIssue::Yes))
                .header("Failed to read downloaded Node.js distribution")
                .body(formatdoc! {"
                    An unexpected I/O error occurred while trying to read the downloaded Node.js \
                    distribution from {path}.
                "})
                .debug_info(source.to_string())
                .create()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use bullet_stream::strip_ansi;
    use heroku_nodejs_utils::package_json::PackageJsonError;
    use insta::{assert_snapshot, with_settings};
    use libherokubuildpack::download::DownloadError;
    use test_support::test_name;

    #[test]
    fn test_inventory_parse_error() {
        assert_error_snapshot(NodeJsEngineBuildpackError::InventoryParse(
            create_toml_error(),
        ));
    }

    #[test]
    fn test_package_json_access_error() {
        assert_error_snapshot(NodeJsEngineBuildpackError::PackageJson(
            PackageJsonError::AccessError(create_io_error("test I/O error blah")),
        ));
    }
    #[test]
    fn test_package_json_parse_error() {
        assert_error_snapshot(NodeJsEngineBuildpackError::PackageJson(
            PackageJsonError::ParseError(create_json_error()),
        ));
    }

    #[test]
    fn test_unknown_version_error() {
        assert_error_snapshot(NodeJsEngineBuildpackError::UnknownVersion(
            "0.0.0".to_string(),
        ));
    }

    #[test]
    fn test_dist_layer_temp_file_error() {
        assert_error_snapshot(DistLayerError::TempFile(create_io_error("Disk full")));
    }

    #[test]
    fn test_dist_layer_download_http_error() {
        assert_error_snapshot(DistLayerError::Download {
            src_url: "https://nodejs.org/download/release/v23.6.0/node-v23.6.0-linux-arm64.tar.gz"
                .to_string(),
            dst_path: "/tmp/some-temp-file".into(),
            source: create_download_http_error(),
        });
    }

    #[test]
    fn test_dist_layer_untar_error() {
        assert_error_snapshot(DistLayerError::Untar {
            src_path: "/tmp/some-temp-file".into(),
            dst_path: "/layers/buildpack/some-layer-dir".into(),
            source: create_io_error("permission denied"),
        });
    }

    #[test]
    fn test_dist_layer_tarball_prefix_error() {
        assert_error_snapshot(DistLayerError::TarballPrefix(
            "https://nodejs.org/download/release/v23.6.0/node-v23.6.0-linux-arm64.tar.gz"
                .to_string(),
        ));
    }

    #[test]
    fn test_dist_layer_installation_error() {
        assert_error_snapshot(DistLayerError::Installation {
            src_path: "/tmp/path-to-src".into(),
            dst_path: "/layers/buildpack/path-to-dst".into(),
            source: create_io_error("unexpected end of file"),
        });
    }

    #[test]
    fn test_dist_layer_checksum_verification_error() {
        assert_error_snapshot(DistLayerError::ChecksumVerification {
            url: "https://nodejs.org/download/release/v23.6.0/node-v23.6.0-linux-arm64.tar.gz"
                .into(),
            expected: "d41d8cd98f00b204e9800998ecf8427e".into(),
            actual: "e62ff0123a74adfc6903d59a449cbdb0".into(),
        });
    }

    #[test]
    fn test_dist_layer_read_temp_file_error() {
        assert_error_snapshot(DistLayerError::ReadTempFile {
            path: "/tmp/path-to-src".into(),
            source: create_io_error("read-only filesystem or storage medium"),
        });
    }

    fn assert_error_snapshot(error: impl Into<Error<NodeJsEngineBuildpackError>>) {
        let error_message = strip_ansi(on_error(error.into()).to_string());
        let test_name = format!(
            "errors__{}",
            test_name()
                .split("::")
                .last()
                .unwrap()
                .trim_start_matches("test")
        );
        with_settings!({
            prepend_module_to_snapshot => false,
            omit_expression => true,
        }, {
            assert_snapshot!(test_name, error_message);
        });
    }

    fn create_io_error(text: &str) -> std::io::Error {
        std::io::Error::new(std::io::ErrorKind::Other, text)
    }

    fn create_toml_error() -> toml::de::Error {
        toml::from_str::<toml::Table>("[[artifacts").unwrap_err()
    }

    fn create_json_error() -> serde_json::error::Error {
        serde_json::from_str::<serde_json::Value>(r#"{\n  "name":\n}"#).unwrap_err()
    }

    fn create_download_http_error() -> DownloadError {
        Box::new(ureq::get("broken/ url").call().unwrap_err()).into()
    }
}
