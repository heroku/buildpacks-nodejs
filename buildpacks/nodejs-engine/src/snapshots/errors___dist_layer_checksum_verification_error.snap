---
source: buildpacks/nodejs-engine/src/errors.rs
---
! Node.js distribution checksum verification failed
!
! An error occurred while verifying the checksum of the Node.js distribution from https://nodejs.org/download/release/v23.6.0/node-v23.6.0-linux-arm64.tar.gz. This error can occur due to an issue with the upstream Node.js download server.
!
! Checksum:
! - Expected: `d41d8cd98f00b204e9800998ecf8427e`
! - Actual: `e62ff0123a74adfc6903d59a449cbdb0`
!
! Use the debug information above to troubleshoot and retry your build.
!
! If the issue persists and you think you found a bug in the buildpack, reproduce the issue locally with a minimal example. Open an issue in the buildpack's GitHub repository and include the details here:
! https://github.com/heroku/buildpacks-nodejs/issues
